builder.py

```diff
@@ -151,101 +142,113 @@ def BuildQueueInitialize(self, queue):
         queue.append('  exit -1')
         queue.append('fi')
-        queue.append('BASE_DIRECTORY=`pwd -P`')
-        queue.append('cd %s || exit $?' % self.projectDefs.GetBuildDirectory())
+
+        # We only need to clean up the repo if we didn't just clone it ...
+        queue.append('if [ $DID_WE_CLONE -eq 0 ]; then')
+        self.BuildQueueCleanup(queue)
+        queue.append('fi')
+
+        # One way or another, we have a clean repository, so get it in a known state:
+        #
+        #   1. git checkout origin/master (in each subproject)
+        #   2. Apply --branch and --subproject as needed
 
         queue.append('')
         queue.append('echo')
-        queue.append('echo ========================= Performing make distclean')
+        queue.append('echo ========================= Performing git checkout origin/master')
         queue.append('date')
-        queue.append('chmod ug+x ./configure; ./configure')
-        queue.append('make distclean')
-        queue.append('echo')
+        queue.append('git checkout origin/master')
+        queue.append('git submodule foreach git checkout origin/master')
 
-        queue.append('')
-        queue.append('if [ ${NEEDS_TFGET} -ne 0 ]; then')
-        queue.append('  echo')
-        queue.append('  echo ========================= Performing tf get')
-        queue.append('  date')
-        queue.append('  tf get')
-        queue.append('  EXITSTATUS=$?')
-        queue.append('  [ $EXITSTATUS != 0 ] && exit $EXITSTATUS')
-        queue.append('  echo')
-        queue.append('fi')
+        if self.config.options.branch:
+            queue.append('')
+            queue.append('echo')
+            queue.append('echo ========================= Performing git checkout %s' % self.config.options.branch)
+            queue.append('date')
+            queue.append('# Applying branch \'%s\' to project' % self.config.options.branch)
+            queue.append('git checkout %s || exit $?' % self.config.options.branch)
 
-        if self.config.options.shelveset:
-            shelvesetList = self.config.options.shelveset.split(',')
+        if self.config.options.subproject:
+            subprojectList = self.config.options.subproject.split(',')
             queue.append('')
             queue.append('echo')
-            queue.append('echo ========================= Performing tf unshelve')
+            queue.append('echo ========================= Performing git checkout')
             queue.append('date')
-            for shelveset in shelvesetList:
-                queue.append('echo \'Unshelving shelveset: %s\'' % shelveset)
-                queue.append('tf unshelve "%s"' % shelveset)
-                queue.append('EXITSTATUS=$?')
-                queue.append('[ $EXITSTATUS != 0 ] && exit $EXITSTATUS')
+            for subproject in subprojectList:
+                # Subproject spec looks like: <dir>:<branch>
+                subprojectElements = subproject.split(':')
```

Comment:
We would not need to split at the semicolon if the input was a dictionary `{ 'folder' : 'branch_name'}`